torch/distributions/constraints.py

```diff
@@ -309,218 +396,337 @@ def check(self, value):
+    each of size `lengths[dim]`, in a way compatible with :func:`torch.cat`.
+    """
+
+    def __init__(
+        self,
+        cseq: Sequence[Con],
+        dim: int = 0,
+        lengths: Optional[Sequence[int]] = None,
+    ) -> None:
+        assert all(isinstance(c, Constraint) for c in cseq)
+        self.cseq: Final[list[Con]] = list(cseq)
+        if lengths is None:
+            lengths = [1] * len(self.cseq)
+        self.lengths: Final[list[int]] = list(lengths)
+        assert len(self.lengths) == len(self.cseq)
+        self.dim: Final[int] = dim
+        super().__init__()
+
+    @property
+    def is_discrete(self) -> bool:
+        return any(c.is_discrete for c in self.cseq)
+
+    @property
+    def event_dim(self) -> int:
+        return max(c.event_dim for c in self.cseq)
+
+    def check(self, value: Tensor) -> Tensor:
+        assert -value.dim() <= self.dim < value.dim()
+        checks = []
+        start = 0
+        for constr, length in zip(self.cseq, self.lengths):
+            v = value.narrow(self.dim, start, length)
+            checks.append(constr.check(v))
+            start = start + length  # avoid += for jit compat
+        return torch.cat(checks, self.dim)
+
 
+class Stack(Constraint, Generic[Con]):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrices at dimension `dim`,
+    in a way compatible with :func:`torch.stack`.
+    """
+
+    def __init__(self, cseq: Sequence[Con], dim: int = 0) -> None:
+        assert all(isinstance(c, Constraint) for c in cseq)
+        self.cseq: Final[list[Con]] = list(cseq)
+        self.dim: Final[int] = dim
+        super().__init__()
 
-class _Boolean(Constraint):
+    @property
+    def is_discrete(self) -> bool:
+        return any(c.is_discrete for c in self.cseq)
+
+    @property
+    def event_dim(self) -> int:
+        dim = max(c.event_dim for c in self.cseq)
+        if self.dim + dim < 0:
+            dim += 1
+        return dim
+
+    def check(self, value: Tensor) -> Tensor:
+        assert -value.dim() <= self.dim < value.dim()
+        vs = [value.select(self.dim, i) for i in range(value.size(self.dim))]
+        return torch.stack(
+            [constr.check(v) for v, constr in zip(vs, self.cseq)], self.dim
+        )
+
+
+# value constraints --------------------------------------------------------------------
+
+
+class Boolean(Constraint):
     """
     Constrain to the two values `{0, 1}`.
     """
 
-    is_discrete = True
+    is_discrete: ClassVar[bool] = True
 
-    def check(self, value):
+    def check(self, value: Tensor) -> Tensor:
         return (value == 0) | (value == 1)
 
 
-class _OneHot(Constraint):
+class Real(Constraint):
     """
-    Constrain to one-hot vectors.
+    Trivially constrain to the extended real line `[-inf, inf]`.
     """
 
-    is_discrete = True
-    event_dim = 1
+    def check(self, value: Tensor) -> Tensor:
+        return value == value.real  # False for NANs.
```

Comment:
Note to self: don't try to fix micro-bugs on the way. I added the `.real` because this check obviously gives a false otherwise for complex-valued arguments, but this causes a CI failure 

> 'aten::real' to ONNX opset version 12 is not supported